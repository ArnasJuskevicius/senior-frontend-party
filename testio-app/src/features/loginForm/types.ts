export interface IForm {
  password: string;
  username: string;
  error: boolean;
}
