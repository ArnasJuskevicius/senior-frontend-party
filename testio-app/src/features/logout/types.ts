export interface IUser {
    isLoginIn: boolean;
}