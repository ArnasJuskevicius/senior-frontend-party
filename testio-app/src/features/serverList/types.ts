export interface ITById<T> {
  [key: string]: T;
}

export interface IList {
  uids: string[];
  byUid: ITById<IListItem>;
  error: boolean;
  isLoading: boolean;
}

export interface IAPIResponse {
  distance: number;
  name: string;
}

export interface IListItem {
  distance: number;
  name: string;
  uid: string;
}
