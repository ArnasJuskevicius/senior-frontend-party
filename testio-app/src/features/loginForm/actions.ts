import axios from 'axios';
import { push } from 'connected-react-router';
import { setLoggedInStatus, setToken } from 'features/user/actions';
import { ITokenAPI } from 'features/user/types';
import { createAction } from 'redux-actions';
import { ThunkAction } from 'redux-thunk';
import { API_ROUTES, APP_ROUTES } from 'Routes';
import { IApp } from 'types';
import {
  REQUEST_FAILED,
  SET_LOGIN_VALUE,
  SET_PASSWORD_VALUE
} from './constants';

export const setLoginInput = createAction(SET_LOGIN_VALUE);
export const setPasswordInput = createAction(SET_PASSWORD_VALUE);
export const setRequestFailed = createAction(REQUEST_FAILED);

// Thunks
export const onSubmit = (): ThunkAction<void, IApp, {}, any> => (
  dispatch,
  getState
) => {
  dispatch(getToken());
};

export const getToken = (): ThunkAction<void, IApp, {}, any> => (
  dispatch,
  getState
) => {
  const state = getState();

  return axios
    .request<ITokenAPI>({
      headers: {
        'content-type': 'application/json'
      },
      method: 'POST',
      params: {
        password: state.form.password,
        username: state.form.username
      },
      url: API_ROUTES.TOKEN
    })
    .then(({ data }) => {
      dispatch(setToken(data.token));
      dispatch(setLoggedInStatus(true));
      dispatch(push(APP_ROUTES.FORM_PAGE));
    })
    .catch(() => {
      dispatch(setRequestFailed(true));
    });
};
