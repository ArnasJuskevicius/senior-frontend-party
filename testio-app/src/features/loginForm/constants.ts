export const SET_LOGIN_VALUE = 'loginForm/SET_LOGIN_VALUE';
export const SET_PASSWORD_VALUE = 'loginForm/SET_PASSWORD_VALUE';
export const REQUEST_FAILED = 'loginForm/REQUEST_FAILED';
export const SET_LOADER = 'loginForm/SET_LOADER';
