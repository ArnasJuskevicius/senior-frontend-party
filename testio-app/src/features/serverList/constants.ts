export const SET_LIST = 'serverList/SET_LIST';
export const SET_ERROR = 'serverList/SET_ERROR';
export const SET_LOADER = 'serverList/SET_LOADER';
