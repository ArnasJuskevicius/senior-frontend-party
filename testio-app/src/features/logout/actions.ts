import { push } from 'connected-react-router'
import { createAction } from 'redux-actions';
import { ThunkAction } from 'redux-thunk';
import { IApp } from '../../types';
import { LOGOUT } from './constants';

export const setLogoutState = createAction(LOGOUT);


export const logout = (): ThunkAction<void, IApp ,{}, any> => (dispatch, getState) => {
    dispatch(setLogoutState(false));
    // remove token and login state from localstorage
    dispatch(push('/'))
};
