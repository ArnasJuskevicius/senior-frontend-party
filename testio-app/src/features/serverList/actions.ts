import axios from 'axios';
import orderBy from 'lodash-es/orderBy';
import { createAction } from 'redux-actions';
import { ThunkAction } from 'redux-thunk';
import { API_ROUTES } from 'Routes';
import { IApp } from 'types';
import { v4 as uuid } from 'uuid';
import { SET_ERROR, SET_LIST } from './constants';
import { IAPIResponse, IListItem } from './types';

export const setServersList = createAction(SET_LIST);
export const setError = createAction(SET_ERROR);

// Thunks
export const getServersList = (): ThunkAction<void, IApp, {}, any> => (
  dispatch,
  getState
) => {
  dispatch(getList());
};

export const buildServerList = (data: IAPIResponse): IListItem => ({
  ...data,
  uid: uuid()
});

export const getList = (): ThunkAction<void, IApp, {}, any> => (
  dispatch,
  getState
) => {
  const { user } = getState();

  return axios
    .request<IAPIResponse[]>({
      headers: {
        Authorization: user.token,
        'content-type': 'application/json'
      },
      method: 'GET',
      url: API_ROUTES.SERVER_LIST
    })
    .then(({ data }) => {
      const sortedArray = orderBy(data, ['distance', 'name']);
      dispatch(setServersList(sortedArray.map(buildServerList)));
    })
    .catch(error => {
      dispatch(setError(true));
    });
};
