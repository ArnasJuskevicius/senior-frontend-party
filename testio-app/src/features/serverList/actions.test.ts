import axios from '__mocks__/axios';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IApp } from 'types';
import { buildServerList, getList } from './actions';
import { SET_ERROR, SET_LIST } from './constants';
import { IAPIResponse } from './types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState: Partial<IApp> = {
  user: {
    isLoggedIn: true,
    token: 'token'
  }
};
let store: any;
const buildPayload = (type: string, payload?: any) => ({
  payload,
  type
});

const testAPIData: IAPIResponse[] = [
  { name: 'Latvia #56', distance: 451 },
  { name: 'Lithuania #56', distance: 1713 },
  { name: 'Latvia #81', distance: 504 }
];

describe('List actions ', () => {
  beforeAll(() => {
    axios.request = jest.fn(() => Promise.resolve({ data: { testAPIData } }));
  });

  beforeEach(() => {
    store = mockStore(initialState);
  });

  describe('#buildAds', () => {
    it('should build countries list from api data', () => {
      const result = testAPIData.map(buildServerList);
      expect(result).toMatchSnapshot();
    });
  });

  describe('#getList', () => {
    it('should dispatch error action then request failed', async () => {
      axios.request.mockImplementationOnce(() => Promise.reject());

      await store.dispatch(getList());
      const actions = store.getActions();

      const expectedPayload = buildPayload(SET_ERROR, true);
      expect(actions).toEqual([expectedPayload]);
    });

    it('should dispatch setServersList action then request succeed', async () => {
      await store.dispatch(getList());
      const actions = store.getActions();
      expect(actions[0].type).toEqual(SET_LIST);
    });
  });
});
