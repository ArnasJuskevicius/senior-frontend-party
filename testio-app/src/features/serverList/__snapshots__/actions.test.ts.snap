// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List actions  #buildAds should build countries list from api data 1`] = `
Array [
  Object {
    "distance": 451,
    "name": "Latvia #56",
    "uid": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "distance": 1713,
    "name": "Lithuania #56",
    "uid": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "distance": 504,
    "name": "Latvia #81",
    "uid": "00000000-0000-0000-0000-000000000000",
  },
]
`;
